### PR TITLE
Change default values of `content`, `content_type` for hash checks

### LINKED ISSUE #24: Requests with missing body or content type cause an unexpected ValueError
**STR:**
Using something like this with web.py:

``` python
    try:
        Receiver(lookup_hawk_credentials,
                 auth_header,
                 url,
                 env.get('REQUEST_METHOD'),
                 content=web.data(),
                 content_type=env.get('CONTENT_TYPE'))
    except HawkFail as e:
        ...
        raise web.Forbidden()
```

...attempt to POST to the API using httpie and requests-hawk:
`http POST localhost:8080/api/saveclassification --auth-type=hawk --auth='treeherder:zz'`

**Expected:**
Either the request succeeds, or `Receiver` raises one of the `HawkFail` class of exceptions, such that `web.Forbidden()` is raised, resulting in an HTTP 403.

**Actual:**

```
HTTP/1.1 500 Internal Server Error
Content-Type: text/html
Date: Wed, 03 Feb 2016 18:27:06 GMT
Server: localhost
Transfer-Encoding: chunked

internal server error
```

And in the server logs:

```
Traceback (most recent call last):
  ...
  File "/home/Ed/src/orangefactor/server/handlers.py", line 504, in ensure_authenticated
    content_type=env.get('CONTENT_TYPE'))
  File "/home/Ed/.virtualenvs/of/lib/python2.7/site-packages/mohawk/receiver.py", line 111, in __init__
    **auth_kw)
  File "/home/Ed/.virtualenvs/of/lib/python2.7/site-packages/mohawk/base.py", line 42, in _authorize
    content_hash = resource.gen_content_hash()
  File "/home/Ed/.virtualenvs/of/lib/python2.7/site-packages/mohawk/base.py", line 203, in gen_content_hash
    'payload content and/or content_type cannot be '
ValueError: payload content and/or content_type cannot be empty without an explicit allowance

127.0.0.1:50832 - - [03/Feb/2016 18:27:06] "HTTP/1.1 POST /api/saveclassification" - 500 Internal Server Error
```

I follow the motivation for trying to be strict about making people validate the content, however:
(a) I'm not sure it's a correct assumption that `None` values means the developer has instantiated `Receiver` incorrectly. I would have thought it's more likely that the request just didn't contain a body or content type to start with.
(b) It would be more consistent to raise a `HawkFail` (or subclass), since it's non-obvious that I should be catching `ValueError` too.

(Using mohawk 0.3.1 with requests-hawk 1.0.0 and httpie 0.9.3 on Python 2.7.11)

### LINKED ISSUE #43: ValueError with accept_untrusted_content=True and hash provided in header.
With a `Receiver` configured to not validate the content-hash

```python
Receiver(lookup_hawk_credentials,
         auth_header,
         url,
         env.get('REQUEST_METHOD'),
         content=None
         content_type=None,
         accept_untrusted_content=True)
```

you get a `ValueError: payload content and/or content_type cannot be empty without an explicit allowance` whenever a client *does* send a content-hash.

The reason is in `HawkAuthority._authorize()`:

```python
       if 'hash' not in parsed_header and accept_untrusted_content:
            # The request did not hash its content.
            log.debug('NOT calculating/verifiying payload hash '
                      '(no hash in header)')
            check_hash = False
            content_hash = None
        else:
            check_hash = True
            content_hash = resource.gen_content_hash()
```

Here the check for the hash supersedes `accept_untrusted_content`.

As stated in #24 I still think these flags add unneeded complexity.

### PR DESCRIPTION
**Pull request by rneilson**
Resolves #24, resolves #43.

* Use new `HawkEmptyValue` class as default values of `content`, `content_type` arguments.
* Raise new `MissingContent` exception (derived from `HawkFail`) when `content` or `content_type` arguments not given and `always_hash_content` not set to `True`.
* Update documentation regarding `accept_untrusted_content` argument to `Receiver` class, clarifying that an omitted content hash is allowed, not that hash check is skipped entirely.

### CODE REVIEW

**Comment by rneilson** [comment]
@kumar303 Okay, decided to roll in a fix for #24 as well as changing the docs on #43 (wasn't much extra effort). Also can't seem to request a review -- I know Github's review functions are newish, but I still haven't quite figured out how to trigger them reliably :(

**Comment by edmorley** [comment]
> Also can't seem to request a review

Unfortunately only repo owners/collaborators can request reviews. I emailed GitHub support requesting that be changed - if you do too perhaps they'll start to raise the priority.

**Comment by rneilson** [comment]
That seems...odd. I'd figured reviews would be even more useful for outside contributors. I'll email GitHub support about that, too.

**Comment by kumar303** [comment]
> Unfortunately only repo owners/collaborators can request reviews.

oh, weird! I didn't know that

File: mohawk/base.py [hunk 1]
```diff
@@ -193,12 +214,12 @@ def content_hash(self):
...
-        if self.content is None or self.content_type is None:
+        if self.content == EmptyValue or self.content_type == EmptyValue:
             if self.always_hash_content:
                 # Be really strict about allowing developers to skip content
                 # hashing. If they get this far they may be unintentiionally
                 # skipping it.
-                raise ValueError(
+                raise MissingContent(
                     'payload content and/or content_type cannot be '
                     'empty without an explicit allowance')
```
Comment (kumar303) [inline mohawk/base.py]: I think this should say "...cannot be empty when always_hash_content is True"

File: mohawk/exc.py [hunk 2]
```diff
@@ -101,3 +101,12 @@ class InvalidBewit(HawkFail):
     The bewit is invalid; e.g. it doesn't contain the right number of
     parameters.
     """
+
+
+class MissingContent(HawkFail):
+    """
+    A payload's `content` or `content_type` were not provided, and
```
Comment (rneilson) [inline mohawk/exc.py]: Done.

File: mohawk/base.py (hunk 1 above)
Comment (rneilson) [inline mohawk/base.py]: Done.

**Comment by rneilson** [comment]
> Is there a test that reproduces #43 ?

Sort of. There's `test_skip_payload_hashing`, and `test_unexpected_unhashed_content`, which test the actual mechanisms. Mostly #43 was a naming problem, which is what I was referring to in #46, and a documentation problem about how to allow skipping hash checks when the client doesn't provide it.

I was wondering about this, though: as checked in `test_unexpected_unhashed_content`, `MisComputedContentHash` is raised, but I'm thinking it should raise `MissingContent` instead. I think [here](https://github.com/rneilson/mohawk/blob/33e7a29ed931bdc58242618ddf9c7d16ef6c7681/mohawk/base.py#L73) would be a good place to check, since we're already testing `if check_hash and not their_hash`.

**Comment by rneilson** [comment]
Sorry, should say #43 is part naming/docs problem, and part `None` vs `''` problem. The latter is what this PR fixes - do you think I should add something like `test_empty_payload_hashing` which uses `None` for `content`/`content_type` but leaves `accept_untrusted_content=False`?

**Comment by rneilson** [comment]
Okay, added this one:

```python
    def test_empty_payload_hashing(self):
        method = 'GET'
        content = None
        content_type = None
        sn = self.Sender(method=method, content=content,
                         content_type=content_type)
        self.receive(sn.request_header, method=method, content=content,
                     content_type=content_type,
                     accept_untrusted_content=True)
```

This tests having `accept_untrusted_content=True` and empty content, but with the sender still including a content hash.

File: mohawk/tests.py [hunk 3]
```diff
@@ -142,29 +143,42 @@ def test_post_content_type_with_trailing_charset(self):
...
+    def test_empty_payload_hashing(self):
+        method = 'GET'
+        content = None
+        content_type = None
+        sn = self.Sender(method=method, content=content,
+                         content_type=content_type)
+        self.assertTrue('hash="' in sn.request_header)
+        self.receive(sn.request_header, method=method, content=content,
+                     content_type=content_type,
+                     accept_untrusted_content=True)
```
Comment (kumar303) [inline mohawk/tests.py]: This test is great, thanks. Can you add another test similar to this that uses `content=None` and `content_type=None` but with `accept_untrusted_content=False`? I think we need to assert that raises an error. I see tests for that using `EmptyValue` but not ones that use `None` values.

File: docs/usage.rst [hunk 4]
```diff
@@ -387,8 +387,14 @@ accept unsigned content with ``accept_untrusted_content=True``:
     ...                     sender.request_header,
```
Comment (kumar303) [inline docs/usage.rst]: While you're in here could you fix the "unsigned" part of the docs? I think it should read more like:

The :class:`mohawk.Receiver` must also be constructed to accept content without a declared hash using accept_untrusted_content=True:

Comment (kumar303) [inline docs/usage.rst]: The sentence above it should be changed as well because it's inaccurate. I think this is better:

In some cases you may not be able to hash request/response content. For example, the content could be too large. If you run into this, Hawk might not be the best fit for you but Hawk does allow you to accept content without a declared hash if you wish.

... important::

  By allowing content without a declared hash, both the sender and receiver are susceptible to content tampering.

File: mohawk/tests.py (hunk 3 above)
Comment (rneilson) [inline mohawk/tests.py]: The thing about changing the default to `EmptyValue` is that `content=None` is now treated the same as `content=''`, so as not to trip on web frameworks that return `None` instead of empty strings, so an error isn't raised in that case anymore. (This is to differentiate between passing in empty content and failing to pass in content at all.)

Comment (kumar303) [inline mohawk/tests.py]: ok, I'd still like to see some tests though because sometimes None can trigger exceptions

File: docs/usage.rst (hunk 4 above)
Comment (rneilson) [inline docs/usage.rst]: Done.

Comment (rneilson) [inline docs/usage.rst]: Done.

File: mohawk/tests.py (hunk 3 above)
Comment (rneilson) [inline mohawk/tests.py]: Added a couple more. So now we test several things: content and type being `None` is now equivalent to them being empty strings, that `always_hash_content=False` will still add the hash even when content and type are `None`, and that `accept_untrusted_content=True` will still check the hash if it's included.

**Comment by rneilson** [comment]
You're quite welcome. Let me know what you think regarding #46 and #47 when you get a chance, too.
